src/python/georeferencer/gcp_generation.py

```diff
@@ -0,0 +1,166 @@
+
+    reshaped_matrix = matrix.reshape(rows // 2, 2, cols // 2, 2)
+    downsampled_matrix = reshaped_matrix.mean(axis=(1, 3))
+
+    return downsampled_matrix
+
+
+def get_variance_array_from_file(path="gcp_library.npz"):
+    """Load the variance array from a specified NPZ file.
+
+    Parameters:
+    path (str): The file path to the NPZ file containing the variance array.
+                Default is 'gcp_library.npz'.
+
+    Returns:
+    np.ndarray: An array of variance values loaded from the NPZ file.
+                The array is expected to be stored under the key 'variance_array'.
+    """
+    return np.load(path, allow_pickle=True)["variance_array"]
+
+
+def save_reference_data(gcp_points, variance_array, path="gcp_library.npz"):
+    """Save GCP (Ground Control Points) and variance array to an NPZ file.
+
+    Parameters:
+    gcp_points (np.ndarray): An array of ground control points to be saved.
+    variance_array (np.ndarray): An array of variance values corresponding to the GCPs.
+    path (str): The file path where the NPZ file will be saved.
+                Default is 'gcp_library.npz'.
+
+    Returns:
+    None: This function does not return any value. It saves the data to the specified file.
+    """
+    np.savez(path, gcp_points=gcp_points, variance_array=variance_array)
+
+
+def get_gcp_points_from_file(path="gcp_library.npz"):
+    """Load GCP (Ground Control Points) from a specified NPZ file.
+
+    Parameters:
+    path (str): The file path to the NPZ file containing the GCP points.
+                Default is 'gcp_library.npz'.
+
+    Returns:
+    np.ndarray: An array of GCP points loaded from the NPZ file.
+                The array is expected to contain the GCP points stored
+                under the key 'gcp_points'.
+    """
+    return np.load(path, allow_pickle=True)["gcp_points"]
+
+
+def get_variance_array(matrix, step=8, box_size=48):
+    """Compute the variance of reflectance values in a matrix.
+
+    Parameters:
+    matrix (np.array, dtype=np.float32): A matrix containing reflectance values
+    step (np.int) : How often the variance is calculated per pixel/line
+    box_size (np.int) : How many pixels in box_size x box_size in which variance is calculated from
+
+    Returns:
+    variance_array (np.array): An array of variances
+    """
+    window_size = (box_size, box_size)
+    height, width = matrix.shape
+    variance_array = np.zeros(((height - box_size) // step + 1, (width - box_size) // step + 1), dtype=np.float32)
+
+    win_mean = ndimage.uniform_filter(matrix, window_size)
+    win_sqr_mean = ndimage.uniform_filter(matrix**2, window_size)
+    variance = win_sqr_mean - win_mean**2
+
+    for i in range(0, height - box_size + 1, step):
+        for j in range(0, width - box_size + 1, step):
+            variance_array[i // step, j // step] = variance[i + box_size // 2][j + box_size // 2]
+    return variance_array
+
+
+def get_gcp_candidates(variance_array, group_size=3):
```

Comment:
Looks like the objective here is to find local maxima, maybe try https://stackoverflow.com/a/47059579 ?